keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,863 @@
+
+#### Add nodes with higher or lower capabilities
+
+A user created a cluster with an ample sized cluster CIDR. All the initial nodes
+are of uniform capacity capable of running a maximum of 256 pods and they are
+each assigned a /24 pod CIDR. The user is planning to add more nodes to the
+system which are capable of running 500 pods. However, they cannot take
+advantage of the additional capacity because all nodes are assigned a /24 pod
+CIDR. With this enhancement the user configures a new allocation which uses the
+original cluster CIDR but allocates a /23 instead of a /24 to each node. They
+use the node selector to allocate these IPs only to the nodes with the higher
+capacity.
+
+#### Provision discontiguous ranges
+
+A user wants to create a cluster with 32 nodes each with a capacity to run 256
+pods. This means that each node needs a /24 pod CIDR range and they need a total
+range of /19. However, there aren't enough contiguous IPs in the user's network.
+They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each
+with a /21 range. The CIDR allocator allocates a /24 range from any of these /21
+ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+A major precondition for this feature is to prevent churn on the IPs assigned to
+Nodes. Specifically, we want to prevent the controller from attempting to change
+a Node’s assigned PodCIDRs. For example, K8s will reject attempts to delete a
+CIDRRange that’s currently in use by a Node. This should prevent accidental
+broad-ranging changes to the cluster.
+
+### Risks and Mitigations
+
+-   Do racing controllers need a more sophisticated design? The current plan is
+    to plug in to the kube-controller-manager leader election.
+-   Can the controller install admission controls? This is a new pattern for
+    in-tree components.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+
+```
+type ClusterCIDRConfig struct {
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    CIDRBlock     string
+
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    *v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
```

Comment:
Yeah, we're allowing overlapping CIDR ranges as part of this proposal. The thought is that we can rely on leader election to prevent overlapping/conflicting allocations to nodes.

Fundamentally, I don't think this is a new problem -- even the existing controller relies on kube-controller-manager's leader election. I don't believe having multiple CIDR blocks makes this problem worse (at least as far as I can see).